# 

In [ ]:
# Import libraries
import pandas as pd
import numpy as np

## Explore log_data

In [ ]:
# Read parsed_log.txt
parsed_log_path = '/work/parsed_log.txt'
df = pd.read_csv(parsed_log_path, sep='\t', encoding='utf-8')

In [ ]:
df.head()

MAC                       SessionMainMenu AppName  LogID  \
0  B046FCAC0DC1  B046FCAC0DC1:2016:02:12:12:35:13:437     VOD     52   
1  B046FCAC0DC1  B046FCAC0DC1:2016:02:11:01:01:56:838    IPTV     40   
2  B046FCAC0DC1  B046FCAC0DC1:2016:02:11:01:02:29:258     VOD     55   
3  B046FCAC0DC1  B046FCAC0DC1:2016:02:12:04:44:59:143    IPTV     18   
4  B046FCAC0DC1  B046FCAC0DC1:2016:02:12:12:35:13:437     VOD     54   

          Event     ItemID  RealTimePlaying Firmware DefaultGateway  \
0       StopVOD  100052388            570.3   2.4.14    192.168.1.1   
1     EnterIPTV        NaN              NaN   2.4.14    192.168.1.1   
2       NextVOD  100052388              NaN   2.4.14    192.168.1.1   
3  ChangeModule        NaN              NaN   2.4.14    192.168.1.1   
4       PlayVOD  100052388              NaN   2.4.14    192.168.1.1   

                     ItemName  ...  LocalType     SubnetMask          IPWan  \
0  Trường Học Moorim (20 Tập)  ...        0.0  255.255.255.0  100.98.91.249   
1                         NaN  ...       77.0  255.255.255.0  100.98.91.249   
2  Trường Học Moorim (20 Tập)  ...       77.0  255.255.255.0  100.98.91.249   
3                         NaN  ...       77.0  255.255.255.0  100.98.91.249   
4  Trường Học Moorim (20 Tập)  ...        0.0  255.255.255.0  100.98.91.249   

  CustomerID   Contract Multicast   PrimaryDNS            IP  \
0   220688.0  SGFD81389       NaN  192.168.1.1  192.168.1.25   
1   220688.0  SGFD81389       NaN  192.168.1.1  192.168.1.25   
2   220688.0  SGFD81389       NaN  192.168.1.1  192.168.1.25   
3   220688.0  SGFD81389       NaN  192.168.1.1  192.168.1.25   
4   220688.0  SGFD81389       NaN  192.168.1.1  192.168.1.25   

                                Session AppID  
0  B046FCAC0DC1:2016:02:12:12:34:51:923   VOD  
1  B046FCAC0DC1:2016:02:11:01:01:56:813  IPTV  
2  B046FCAC0DC1:2016:02:11:01:01:56:813   VOD  
3  B046FCAC0DC1:2016:02:12:04:44:54:688   VOD  
4  B046FCAC0DC1:2016:02:12:12:34:51:923   VOD  

[5 rows x 24 columns]

In [ ]:
# 1. Get a summary of the DataFrame
summary = df.info()

# 3. Examine summary statistics for numeric columns
num_stats = df.describe()

# 4. Check the number and names of unique values in each column
unique_counts = df.nunique()

summary, num_stats, unique_counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 914060 entries, 0 to 914059
Data columns (total 24 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   MAC              914060 non-null  object 
 1   SessionMainMenu  914032 non-null  object 
 2   AppName          914060 non-null  object 
 3   LogID            914060 non-null  int64  
 4   Event            914060 non-null  object 
 5   ItemID           760938 non-null  object 
 6   RealTimePlaying  294031 non-null  float64
 7   Firmware         914060 non-null  object 
 8   DefaultGateway   914060 non-null  object 
 9   ItemName         757649 non-null  object 
 10  IsLandingPage    571834 non-null  float64
 11  BoxTime          914060 non-null  object 
 12  Screen           792552 non-null  object 
 13  SecondaryDNS     914060 non-null  object 
 14  LocalType        907408 non-null  float64
 15  SubnetMask       914060 non-null  object 
 16  IPWan            914060 non-null  obje

(None,
                LogID  RealTimePlaying  IsLandingPage      LocalType  \
 count  914060.000000     2.940310e+05  571834.000000  907408.000000   
 mean       78.748688     1.237851e+03       0.112547       5.442194   
 std       122.856225     6.018782e+04       0.316038      24.711289   
 min        12.000000     1.500000e-01       0.000000       0.000000   
 25%        41.000000     5.713000e+00       0.000000       0.000000   
 50%        42.000000     1.628600e+01       0.000000       0.000000   
 75%        51.000000     2.557925e+02       0.000000       0.000000   
 max       512.000000     1.743658e+07       1.000000     197.000000   
 
          CustomerID  
 count  9.040960e+05  
 mean   6.528100e+05  
 std    1.786735e+06  
 min    8.091500e+04  
 25%    2.329170e+05  
 50%    3.646910e+05  
 75%    5.392930e+05  
 max    1.000026e+08  ,
 MAC                  1693
 SessionMainMenu    101190
 AppName                 2
 LogID                  34
 Event                  34


From the exploratory data analysis, summary of the dataframe:

- The DataFrame 'df' has 914060 entries and 24 columns.
- No column is entirely null, but columns such as SessionMainMenu, ItemID, RealTimePlaying, ItemName, IsLandingPage, Screen, LocalType, CustomerID, Contract, and Multicast have some null entries.
- The DataFrame has a combination of object, int64, and float64 data types.
- It appears there are 20885 duplicate entries in the DataFrame.
- Summary statistics for numeric columns provide a snapshot of the data distribution for those columns. For example, the 'LogID' column has a mean of ~78.75 with a standard deviation of ~122.85. The maximum value is 512, and the minimum value is 12. Similar insights can be drawn for other numeric columns.
- Unique counts gives us a sense of the data cardinality. For instance, there are 1693 unique values in the 'MAC' column, and only 2 unique values in the 'AppName' column.


In [ ]:
df = df.drop(columns=['Firmware', 'DefaultGateway', 'IsLandingPage', 'Screen', 'SecondaryDNS', 'LocalType', 'SubnetMask', 'IPWan', 'Multicast','CustomerID','PrimaryDNS', 'IP'])

In [ ]:
# Check for duplicates
df[df.duplicated()]

MAC                       SessionMainMenu AppName  LogID  \
54616   B046FCA981B6  B046FCA981B6:2016:02:03:18:20:14:926     VOD    512   
54617   B046FCA981B6  B046FCA981B6:2016:02:03:10:35:52:709    IPTV     41   
54618   B046FCA981B6  B046FCA981B6:2016:02:03:18:48:25:838    IPTV    411   
54619   B046FCA981B6  B046FCA981B6:2016:02:04:19:10:09:438    IPTV     41   
54620   B046FCA981B6  B046FCA981B6:2016:02:04:19:10:09:438    IPTV     41   
...              ...                                   ...     ...    ...   
914055  B046FCAD1DE8  B046FCAD1DE8:2016:03:11:14:34:33:670    IPTV     42   
914056  B046FCAD1DE8   B046FCAD1DE8:2016:03:11:18:41:56:10     VOD     57   
914057  B046FCAD1DE8   B046FCAD1DE8:2016:03:11:18:41:56:10     VOD     57   
914058  B046FCAD1DE8  B046FCAD1DE8:2016:03:11:14:34:07:196     VOD     18   
914059  B046FCAD1DE8   B046FCAD1DE8:2016:03:11:18:41:56:10     VOD     53   

                      Event     ItemID  RealTimePlaying            ItemName  \
54616        EnterDetailVOD  100045181              NaN            Ngày Tàn   
54617          StartChannel         38              NaN         VĨNH LONG 1   
54618   GetDRMKeySuccessful        NaN              NaN                 NaN   
54619          StartChannel        410              NaN             YÊN BÁI   
54620          StartChannel        231              NaN           QUẢNG TRỊ   
...                     ...        ...              ...                 ...   
914055          StopChannel        175         1412.898  ASIAN FOOD CHANNEL   
914056       EnterFolderVOD          1              NaN            MỚI NHẤT   
914057       EnterFolderVOD          1              NaN            MỚI NHẤT   
914058         ChangeModule        NaN              NaN                 NaN   
914059             PauseVOD  100054410              NaN          Diệp Vấn 3   

                        BoxTime   Contract  \
54616   2016:02:03:18:20:19:462  TVDN00002   
54617   2016:02:03:10:38:19:893  TVDN00002   
54618   2016:02:03:18:48:35:459  TVDN00002   
54619   2016:02:04:22:03:35:326  TVDN00002   
54620   2016:02:04:22:10:18:631  TVDN00002   
...                         ...        ...   
914055  2016:03:11:14:58:54:786  SGFDN0084   
914056  2016:03:11:18:41:57:314  SGFDN0084   
914057  2016:03:11:19:14:00:215  SGFDN0084   
914058   2016:03:11:14:34:32:74  SGFDN0084   
914059  2016:03:11:18:43:09:869  SGFDN0084   

                                     Session AppID  
54616   B046FCA981B6:2016:02:03:16:36:19:792   VOD  
54617   B046FCA981B6:2016:02:03:10:35:52:684  IPTV  
54618   B046FCA981B6:2016:02:03:18:48:21:349  IPTV  
54619   B046FCA981B6:2016:02:04:19:10:04:876  IPTV  
54620   B046FCA981B6:2016:02:04:19:10:04:876  IPTV  
...                                      ...   ...  
914055  B046FCAD1DE8:2016:03:11:11:26:51:997  IPTV  
914056  B046FCAD1DE8:2016:03:11:15:34:08:840   VOD  
914057  B046FCAD1DE8:2016:03:11:15:34:08:840   VOD  
914058  B046FCAD1DE8:2016:03:11:11:26:51:997  IPTV  
914059  B046FCAD1DE8:2016:03:11:15:34:08:840   VOD  

[20885 rows x 12 columns]

In [ ]:
# Drop duplicates in the Dataset
df=df.drop_duplicates()

In [ ]:
# Replace missing values in 'RealTimePlaying' with 0
df['RealTimePlaying'] = df['RealTimePlaying'].fillna(0)

In [ ]:
# Replace missing values in 'SessionMainMenu' and 'ItemID' with 'NA'
df['SessionMainMenu'] = df['SessionMainMenu'].fillna('NA')
df['ItemID'] = df['ItemID'].fillna('NA')

In [ ]:
# Extract datetime from 'SessionMainMenu'
df['SessionMainMenu'] = df['SessionMainMenu'].apply(lambda x: x.split(':')[1:] if isinstance(x, str) else x)
df['SessionMainMenu'] = df['SessionMainMenu'].apply(lambda x: ':'.join(x) if isinstance(x, list) else x)

# Convert 'SessionMainMenu' column to datetime
df['SessionMainMenu'] = pd.to_datetime(df['SessionMainMenu'], format='%Y:%m:%d:%H:%M:%S:%f', errors='coerce')


I have replaced the SessionMainMenu column value with only its date and time part and then converted it into datetime format. Now SessionMainMenu column contains only the datetime values.

In [ ]:
# Check for NaN values
df.isna().sum()

MAC                     0
SessionMainMenu        28
AppName                 0
LogID                   0
Event                   0
ItemID                  0
RealTimePlaying         0
ItemName           150558
BoxTime                 0
Contract             5892
Session                 0
AppID                   0
dtype: int64

In [ ]:
# Count of unique values for each column
unique_counts = df.nunique()
unique_counts

MAC                  1693
SessionMainMenu    101185
AppName                 2
LogID                  34
Event                  34
ItemID               4471
RealTimePlaying    114040
ItemName             5117
BoxTime            892962
Contract             1825
Session             66272
AppID                   7
dtype: int64

In [ ]:
df.head()

MAC         SessionMainMenu AppName  LogID         Event  \
0  B046FCAC0DC1 2016-02-12 12:35:13.437     VOD     52       StopVOD   
1  B046FCAC0DC1 2016-02-11 01:01:56.838    IPTV     40     EnterIPTV   
2  B046FCAC0DC1 2016-02-11 01:02:29.258     VOD     55       NextVOD   
3  B046FCAC0DC1 2016-02-12 04:44:59.143    IPTV     18  ChangeModule   
4  B046FCAC0DC1 2016-02-12 12:35:13.437     VOD     54       PlayVOD   

      ItemID  RealTimePlaying                    ItemName  \
0  100052388            570.3  Trường Học Moorim (20 Tập)   
1         NA              0.0                         NaN   
2  100052388              0.0  Trường Học Moorim (20 Tập)   
3         NA              0.0                         NaN   
4  100052388              0.0  Trường Học Moorim (20 Tập)   

                   BoxTime   Contract                               Session  \
0  2016:02:12:12:46:23:663  SGFD81389  B046FCAC0DC1:2016:02:12:12:34:51:923   
1  2016:02:11:01:02:09:375  SGFD81389  B046FCAC0DC1:2016:02:11:01:01:56:813   
2  2016:02:11:03:06:16:654  SGFD81389  B046FCAC0DC1:2016:02:11:01:01:56:813   
3  2016:02:12:04:45:40:984  SGFD81389  B046FCAC0DC1:2016:02:12:04:44:54:688   
4  2016:02:12:12:47:37:208  SGFD81389  B046FCAC0DC1:2016:02:12:12:34:51:923   

  AppID  
0   VOD  
1  IPTV  
2   VOD  
3   VOD  
4   VOD

## Explore user_data

In [ ]:
# Read parsed_log.txt
user_info = pd.read_csv('/work/user_info.txt', sep='\t', encoding='utf-8')

In [ ]:
# Remove 'FBOX' prefix from the MAC column
user_info['MAC'] = user_info['MAC'].str.replace('FBOX', '')

In [ ]:
user_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2692 entries, 0 to 2691
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   MAC        2692 non-null   object
 1   # of days  2692 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 42.2+ KB


In [ ]:
# Check for duplicates
user_info[user_info.duplicated()]

MAC  # of days
237   B046FCA981B6          0
1899  B046FCA6B160          0
2441  B046FCA6B160          4
2551  B046FCA6B160          0
2554  B046FCA6B160          0
2558  B046FCB38384          0
2581  B046FCB38384          0

In [ ]:
# Remove duplicates MAC and keep the row with the highest '# of days'
user_info = user_info.sort_values("# of days", ascending=False).drop_duplicates('MAC')

user_info.describe()

# of days
count  2664.000000
mean    282.105105
std     201.231987
min       0.000000
25%     120.750000
50%     267.000000
75%     411.250000
max    1983.000000

## Merge data

In [ ]:
# Merge user_data with log_data
HDdata = pd.merge(df, user_info, on='MAC', how='inner')
HDdata

MAC         SessionMainMenu AppName  LogID            Event  \
0       B046FCAC0DC1 2016-02-12 12:35:13.437     VOD     52          StopVOD   
1       B046FCAC0DC1 2016-02-11 01:01:56.838    IPTV     40        EnterIPTV   
2       B046FCAC0DC1 2016-02-11 01:02:29.258     VOD     55          NextVOD   
3       B046FCAC0DC1 2016-02-12 04:44:59.143    IPTV     18     ChangeModule   
4       B046FCAC0DC1 2016-02-12 12:35:13.437     VOD     54          PlayVOD   
...              ...                     ...     ...    ...              ...   
893170  B046FCB626E5 2016-03-14 12:18:04.612     VOD     57   EnterFolderVOD   
893171  B046FCB626E5 2016-03-14 12:15:19.629    IPTV    451  ExitChannelList   
893172  B046FCB626E5 2016-03-14 12:45:57.625    IPTV     42      StopChannel   
893173  B046FCB626E5 2016-03-14 17:34:06.842    IPTV     45  ShowChannelList   
893174  B046FCB626E5 2016-03-14 17:34:06.842    IPTV     42      StopChannel   

           ItemID  RealTimePlaying                    ItemName  \
0       100052388          570.300  Trường Học Moorim (20 Tập)   
1              NA            0.000                         NaN   
2       100052388            0.000  Trường Học Moorim (20 Tập)   
3              NA            0.000                         NaN   
4       100052388            0.000  Trường Học Moorim (20 Tập)   
...           ...              ...                         ...   
893170          4            0.000                   HÀNH ĐỘNG   
893171          1            6.820             TRUYỀN HÌNH FPT   
893172          1           24.418                        VTV1   
893173          1            0.000                        VTV1   
893174          1            9.986                        VTV1   

                        BoxTime   Contract  \
0       2016:02:12:12:46:23:663  SGFD81389   
1       2016:02:11:01:02:09:375  SGFD81389   
2       2016:02:11:03:06:16:654  SGFD81389   
3       2016:02:12:04:45:40:984  SGFD81389   
4       2016:02:12:12:47:37:208  SGFD81389   
...                         ...        ...   
893170  2016:03:14:12:18:24:966  DNFD26951   
893171  2016:03:14:12:17:30:895  DNFD26951   
893172  2016:03:14:12:46:25:922  DNFD26951   
893173  2016:03:14:17:35:05:771  DNFD26951   
893174  2016:03:14:17:34:20:914  DNFD26951   

                                     Session AppID  # of days  
0       B046FCAC0DC1:2016:02:12:12:34:51:923   VOD        375  
1       B046FCAC0DC1:2016:02:11:01:01:56:813  IPTV        375  
2       B046FCAC0DC1:2016:02:11:01:01:56:813   VOD        375  
3       B046FCAC0DC1:2016:02:12:04:44:54:688   VOD        375  
4       B046FCAC0DC1:2016:02:12:12:34:51:923   VOD        375  
...                                      ...   ...        ...  
893170  B046FCB626E5:2016:03:14:12:15:19:602   VOD         52  
893171  B046FCB626E5:2016:03:14:12:15:19:602  IPTV         52  
893172  B046FCB626E5:2016:03:14:12:45:50:179  IPTV         52  
893173  B046FCB626E5:2016:03:14:12:55:55:215  IPTV         52  
893174  B046FCB626E5:2016:03:14:12:55:55:215  IPTV         52  

[893175 rows x 13 columns]

In [ ]:
# Save the combined data
HDdata.to_csv("HDdata.csv", index=False)